### PR TITLE
PHP 8.1 / 8.2 Compatibility

### PR DESCRIPTION
**Pull request by swissspidy**
## Summary

Since we use parts of the AMP plugin in the Web Stories plugin, it would be reassuring for us if it works more or less seamlessly with PHP 8.1 and 8.2.

PHP 8.2 compatibility is currently blocked by work on WordPress core itself.

Even PHP 8.1 is not seamless due to the conflict with the Requests library.

See also https://github.com/ampproject/amp-toolbox-php/pull/533 (for which an update to `dev-main` is included in this PR)

## Checklist

- [ ] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [ ] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

File: amp.php
Comment (westonruter) [inline amp.php]: Can be removed now that PHP files have changed.

```suggestion
```

**Comment by swissspidy** [comment]
Something to fix in PHP-CSS-Parser:

```
PHP Deprecated:  Using ${var} in strings is deprecated, use {$var} instead in /tmp/wordpress/src/wp-content/plugins/amp/vendor/sabberworm/php-css-parser/src/Parsing/ParserState.php on line 129
PHP Deprecated:  Using ${var} in strings is deprecated, use {$var} instead in /tmp/wordpress/src/wp-content/plugins/amp/vendor/sabberworm/php-css-parser/src/Parsing/ParserState.php on line 129
PHP Deprecated:  Using ${var} in strings is deprecated, use {$var} instead in /tmp/wordpress/src/wp-content/plugins/amp/vendor/sabberworm/php-css-parser/src/Parsing/ParserState.php on line 130
PHP Deprecated:  Using ${var} in strings is deprecated, use {$var} instead in /tmp/wordpress/src/wp-content/plugins/amp/vendor/sabberworm/php-css-parser/src/Parsing/ParserState.php on line 131
```

File: .github/workflows/build-test-measure.yml
Comment (thelovekesh) [inline .github/workflows/build-test-measure.yml:499]: I think pcov is disabled temporarily, right?

Comment (swissspidy) [reply .github/workflows/build-test-measure.yml:499]: PCOV support was merged into PHPUnit 8. `pcov/clobber` is no longer needed.

File: tests/php/src/Instrumentation/EventTest.php
Comment (westonruter) [inline tests/php/src/Instrumentation/EventTest.php:87]: See https://github.com/ampproject/amp-wp/commit/924be547674068a54c62fb87fcc207f03f6af27f

> Using floats as array keys is deprecated in PHP 8.1

File: tests/php/bootstrap.php
Comment (westonruter) [inline tests/php/bootstrap.php:50]: Why was the `file_exists()` check needed? Is it due to WP-CLI not being able to be installed on PHP 8.2?

File: includes/validation/class-amp-validation-callback-wrapper.php
Comment (westonruter) [inline includes/validation/class-amp-validation-callback-wrapper.php:396]: I suppose this is needed because we can't specify the matching return type since return types aren't available in older versions of PHP. 

Oh, but actually we now require PHP 7 in the AMP plugin. So instead of this attribute could we instead just add the return types?

Comment (swissspidy) [reply includes/validation/class-amp-validation-callback-wrapper.php:396]: If we‘re not worried about someone extending the class and breaking their implementation because of the return types, sure.

**Comment by westonruter** [comment]
There's currently an error for PHP 7.4 with coverage.

File: tests/php/bootstrap.php
Comment (swissspidy) [reply tests/php/bootstrap.php:50]: Not 100% sure right now (not at computer) but yeah I think it was because of issues installing it. Maybe because of Requests? Will have to check it out again.

File: includes/validation/class-amp-validation-callback-wrapper.php
Comment (swissspidy) [reply includes/validation/class-amp-validation-callback-wrapper.php:396]: See https://php.watch/versions/8.1/ReturnTypeWillChange

File: phpunit.xml.dist
Comment (swissspidy) [inline phpunit.xml.dist:8]: This can be reverted again once there are no warnings from WP core anymore (looking at you, Requests library)

File: src/Dom/ElementList.php
Comment (swissspidy) [inline src/Dom/ElementList.php]: I think here we can safely add the return types because it's a final class

File: tests/php/bootstrap.php
Comment (swissspidy) [reply tests/php/bootstrap.php:50]: Here's the run before that change: https://github.com/ampproject/amp-wp/actions/runs/2925960305/jobs/4667211482

```
PHP Warning:  require_once(/tmp/wordpress/src/wp-content/plugins/amp/vendor/wp-cli/wp-cli/php/utils.php): Failed to open stream: No such file or directory in /tmp/wordpress/src/wp-content/plugins/amp/tests/php/bootstrap.php on line 62
```

**Comment by swissspidy** [comment]
> There's currently an error for PHP 7.4 with coverage.

Yeah I saw it too and it's pretty confusing.

Somehow the test never finishes, so there's no coverage file being created:

https://github.com/ampproject/amp-wp/actions/runs/3087716934/jobs/5021640946#step:21:55

A couple of PHP warnings:

```
PHP Warning:  Cannot modify header information - headers already sent by (output started at /tmp/wordpress-tests-lib/includes/bootstrap.php:261) in /tmp/wordpress/src/wp-includes/pluggable.php on line 1421

Warning: Cannot modify header information - headers already sent by (output started at /tmp/wordpress-tests-lib/includes/bootstrap.php:261) in /tmp/wordpress/src/wp-includes/pluggable.php on line 1424
```

Then I noticed that while CI is supposed to use PCOV for coverage ever since https://github.com/ampproject/amp-wp/commit/536dbe176596c375becaeadd1151833d9c7353e9, it has always been using XDebug.

See https://github.com/ampproject/amp-wp/actions/runs/2749592396/jobs/4314380797#step:20:17 from 2 months ago.

Note how it says "Runtime:       PHP 7.4.30 with Xdebug"

So it was never using PCOV at all.

Now I've tried switching to XDebug again but it doesn't work either. Same error though 😞 

No idea why. `wp-includes/pluggable.php on line 1424` is the `wp_redirect()` function. So something is doing redirects when trying to install WordPress...

File: .phpcs.xml.dist
Comment (swissspidy) [inline .phpcs.xml.dist:115]: The plugin requires PHP 7+

Without this we're getting yelled at by PHPCS for using return type declarations.

Comment (westonruter) [reply .phpcs.xml.dist:115]: Oh, good catch. This should have been updated as part of #7088.

File: includes/validation/class-amp-validation-callback-wrapper.php
Comment (westonruter) [reply includes/validation/class-amp-validation-callback-wrapper.php:396]: No, I'm not worried about someone extending the class. It should have been marked `final` to begin with.

Comment (swissspidy) [reply includes/validation/class-amp-validation-callback-wrapper.php:396]: Since we can't use `void` return types yet (they're a PHP 7.1 feature), can't really change this.

**Comment by swissspidy** [comment]
@westonruter I found the issue 🥳 🎉 

The culprit was the `test_handle_single_url_page_bulk_and_inline_actions` test covering `\AMP_Validation_Error_Taxonomy::handle_single_url_page_bulk_and_inline_actions()`. That method calls `wp_safe_redirect` and `exit()`, causing the tests to silently halt.

I update the test to short-circuit the redirect and wrapping the `exit` in an `if` so that it's never called during tests.

While at it, I noticed that `test_prepare_response_throwing_error` failed on trunk because on trunk there are currently tons of other PHP deprecation warnings, making the regex not match correctly. Making the regex less strict resolves the issue.

All tests should be passing now.

**Comment by swissspidy** [comment]
Slight correction, there are two more tests failing on trunk. They are unrelated to the changes here though.

```
1) AMP_Style_Sanitizer_Test::test_prioritized_stylesheets with data set "admin_bar_included" (Closure Object (...), Closure Object (...))
Failed asserting that '' contains "admin-bar".

Error: /tmp/wordpress/src/wp-content/plugins/amp/tests/php/test-amp-style-sanitizer.php:3494
/tmp/wordpress/src/wp-content/plugins/amp/tests/php/test-amp-style-sanitizer.php:3575

2) Test_AMP_Theme_Support::test_prepare_response_standard
'<link rel="preload" as="script" href="https://cdn.ampproject.org/v0/amp-dynamic-css-classes-0.1.js">' is not after '<link rel="dns-prefetch" href="//cdn.ampproject.org">'
Failed asserting that 364 is greater than 61757.

/tmp/wordpress/src/wp-content/plugins/amp/tests/php/test-class-amp-theme-support.php:1767

```

Both fail because the HTML output is polluted with deprecation warnings like this from core:

```
Deprecated: strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated in /tmp/wordpress/wp-content/plugins/gutenberg/lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php on line 1113
```

This bug was just fixed in Gutenberg yesterday: https://github.com/WordPress/gutenberg/pull/44829
So this will get resorted once the next GB version is released.

File: includes/validation/class-amp-validation-manager.php
Comment (swissspidy) [inline includes/validation/class-amp-validation-manager.php:275]: Prevents deprecation warnings for automatic conversion of `false` to `array`

File: .github/workflows/build-test-measure.yml
Comment (thelovekesh) [inline .github/workflows/build-test-measure.yml:437]: Should we continue using `pcov` for coverage in the CI? I've looked at the problem that it uses xdebug in any case, but I can debug it in another PR. Thoughts?

File: composer.json
Comment (thelovekesh) [inline composer.json:17]: @westonruter Is it intended to bump it to `dev-main`?

File: phpunit.xml.dist
Comment (thelovekesh) [inline phpunit.xml.dist:8]: It can be reverted now?

**Review by thelovekesh — APPROVED** [review]
Aside from the tests that failed and will be fixed following the new GB release, the rest LGTM 👍🏼

File: .github/workflows/build-test-measure.yml
Comment (swissspidy) [reply .github/workflows/build-test-measure.yml:437]: I'll defer to Weston, but if it works and is faster, sure.

But since this never really worked in the first place, definitely warrants its own issue/PR.

File: phpunit.xml.dist
Comment (swissspidy) [reply phpunit.xml.dist:8]: Not yet. See https://github.com/ampproject/amp-wp/pull/7225#discussion_r977290405

File: .github/workflows/build-test-measure.yml
Comment (thelovekesh) [reply .github/workflows/build-test-measure.yml:437]: PCOV is faster but lacks path coverage. A difference in reporting can be seen [here](https://github.com/krakjoe/pcov#differences-in-reporting).

File: composer.json
Comment (westonruter) [reply composer.json:17]: Yes, in order to update php-toolbox-php to the version that adds PHP 8.2 support: https://github.com/ampproject/amp-toolbox-php/pull/533

**Comment by swissspidy** [comment]
> This bug was just fixed in Gutenberg yesterday: [WordPress/gutenberg#44829](https://github.com/WordPress/gutenberg/pull/44829)
> So this will get resorted once the next GB version is released.

FYI, while GB 14.4 was released _today_, that bugfix won't be included until version 14.5 ([due by November 07, 2022](https://github.com/WordPress/gutenberg/milestone/190))

**Comment by swissspidy** [comment]
> Unit test (with coverage): PHP 7.4, WP latest

This required status check needs to be removed/changed, since now the coverage is done with PHP 8.0 instead.

**Comment by westonruter** [comment]
> > Unit test (with coverage): PHP 7.4, WP latest
> 
> This required status check needs to be removed/changed, since now the coverage is done with PHP 8.0 instead.

Removed (to be changed after merge)

File: .github/workflows/build-test-measure.yml
Comment (westonruter) [reply .github/workflows/build-test-measure.yml:437]: PCOV was introduced in https://github.com/ampproject/amp-wp/pull/5201

I just pinged you both the thread where this was discussed.

File: includes/validation/class-amp-validation-manager.php
Comment (westonruter) [inline includes/validation/class-amp-validation-manager.php]: Isn't the return value now always made an array?

```suggestion
	 * @return array Filtered options.
```

File: includes/validation/class-amp-validation-error-taxonomy.php
Comment (westonruter) [inline includes/validation/class-amp-validation-error-taxonomy.php]:
```suggestion
			exit(); // @codeCoverageIgnore
```

File: src/Admin/SiteHealth.php
Comment (westonruter) [inline src/Admin/SiteHealth.php:543]: This was introduced in bd2ee42bf5c035ad1d3868d6f40400d6af13d99d apparently to deal with a PHPCS bug, and not due to any fault of the code?

File: tests/php/bootstrap.php
Comment (westonruter) [inline tests/php/bootstrap.php:30]: Nice catch

File: src/Admin/SiteHealth.php
Comment (swissspidy) [reply src/Admin/SiteHealth.php:543]: Correct

**Review by westonruter — APPROVED** [review]
Thanks for doing this! 

Shall we wait to merge until Gutenberg 14.5 is released?

**Comment by swissspidy** [comment]
I don‘t see an issue with merging this already, but up to you

**Comment by westonruter** [comment]
The issue is the failing tests will make it hard to identify problems with other PRs being opened in the next ~2 weeks.

**Comment by swissspidy** [comment]
You can also just temporarily skip these two tests

**Comment by westonruter** [comment]
Gutenberg updated in https://github.com/ampproject/amp-wp/pull/7331. PHPUnit tests now all pass. 🎉

**Comment by pavanpatil1** [comment]
Tested it with PHP 8.1 and 8.2, Plugin functionality seems working fine with both the PHP versions. However, there are some deprecation warnings are visible in both versions. 

**PHP8.1**

```php
PHP Deprecated:  header(): Passing null to parameter #3 ($response_code) of type int is deprecated in C:\Users\win 10\Local Sites\php8amp\app\public\wp-content\plugins\amp\includes\class-amp-http.php on line 95
```


**PHP8.2**

```php
PHP Deprecated:  Return type of AMP_Validation_Callback_Wrapper::offsetUnset($offset) should either be compatible with ArrayAccess::offsetUnset(mixed $offset): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /var/www/htdocs/current/wp-content/plugins/amp/includes/validation/class-amp-validation-callback-wrapper.php on line 425
PHP Deprecated:  Creation of dynamic property AmpProject\Optimizer\Configuration\MinifyHtmlConfiguration::$minifyAmpScript is deprecated in /var/www/htdocs/current/wp-content/plugins/amp/vendor/ampproject/amp-toolbox/src/Optimizer/Configuration/BaseTransformerConfiguration.php on line 40
PHP Deprecated:  header(): Passing null to parameter #3 ($response_code) of type int is deprecated in /var/www/htdocs/current/wp-content/plugins/amp/includes/class-amp-http.php on line 92
```

**Comment by westonruter** [comment]
In regards to:

```
PHP Deprecated:  header(): Passing null to parameter #3 ($response_code) of type int is deprecated in C:\Users\win 10\Local Sites\php8amp\app\public\wp-content\plugins\amp\includes\class-amp-http.php on line 95
```

This is the code in question:

https://github.com/ampproject/amp-wp/blob/68d9bb3f977967fa05fb09f5f52da561dc85bd72/includes/class-amp-http.php#L78-L98

It looks like all that is required is:

```diff
- 				'status_code' => null,
+ 				'status_code' => 0,
```

**Comment by westonruter** [comment]
In regards to:

```
PHP Deprecated:  Return type of AMP_Validation_Callback_Wrapper::offsetUnset($offset) should either be compatible with ArrayAccess::offsetUnset(mixed $offset): void, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in /var/www/htdocs/current/wp-content/plugins/amp/includes/validation/class-amp-validation-callback-wrapper.php on line 425
```

I'm confused since it _does_ have `#[\ReturnTypeWillChange]`:

https://github.com/ampproject/amp-wp/blob/68d9bb3f977967fa05fb09f5f52da561dc85bd72/includes/validation/class-amp-validation-callback-wrapper.php#L422-L433

**Comment by westonruter** [comment]
And finally, in regards to:

```
PHP Deprecated:  Creation of dynamic property AmpProject\Optimizer\Configuration\MinifyHtmlConfiguration::$minifyAmpScript is deprecated in /var/www/htdocs/current/wp-content/plugins/amp/vendor/ampproject/amp-toolbox/src/Optimizer/Configuration/BaseTransformerConfiguration.php on line 40
```

Looks like this is something that was missed in https://github.com/ampproject/amp-toolbox-php/pull/533. 

See example at https://github.com/ampproject/amp-toolbox-php/pull/533/files#diff-12262a92eb1759cfc297901a94d3e5ab9edc9862154fb003e21f7b80719a7fc2

**Comment by westonruter** [comment]
Actually, it seems like this was tested without the latest `amp-toolbox-php`. Line 40 in `BaseTransformerConfiguration.php` is the [following in v0.11.2](https://github.com/ampproject/amp-toolbox-php/blob/8fa158845c980437831397100ca7133eee76c61d/src/Optimizer/Configuration/BaseTransformerConfiguration.php#L40):

```php
$this->$key = $this->validate($key, $value);
```

Whereas in `main` it is the [following](https://github.com/ampproject/amp-toolbox-php/blob/bd739ddcc839a70634bfefe012cbcd2268b51f7e/src/Optimizer/Configuration/BaseTransformerConfiguration.php#L40):

```php
$this->allowedKeys = $this->getAllowedKeys();
```

I just tried downloading the most recent development build and I see that the former is present as expected.

@pavanpatil1 Could you please re-test, making sure you have the latest [development build](https://github.com/ampproject/amp-wp/wiki/Development-Builds) installed (as for all current QA)?

**Comment by thelovekesh** [comment]
I checked with PHP 8.2 but was unable to get any deprecation warnings other than the
```bash
[06-Feb-2023 19:38:52 UTC] PHP Deprecated:  header(): Passing null to parameter #3 ($response_code) of type int is deprecated in /app/public/content/plugins/amp/includes/class-amp-http.php on line 92
```

**Comment by pavanpatil1** [comment]
QA passed ✅ 

Tested it with the latest [development build](https://storage.googleapis.com/ampwp_github_artifacts/refs/heads/develop/dev/amp.zip). Functionality was working as expected before now the deprecation warnings are not visible now with both v8.1 and 8.2. solved in [this](https://github.com/ampproject/amp-wp/pull/7452) PR.

**Pull request merged**
